runtime/compiler/optimizer/TreeLowering.cpp

```diff
@@ -719,3 +756,719 @@ TR::TreeLowering::lowerArrayStoreCHK(TR::Node *node, TR::TreeTop *tt)
+   TR::Node *anchoredElementIndexNode = anchoredElementIndexTT->getNode()->getFirstChild();
+
+   TR::Node *elementAddress = J9::TransformUtil::calculateElementAddress(comp, anchoredArrayBaseAddressNode, anchoredElementIndexNode, TR::Address);
+   TR::SymbolReference *elementSymRef = comp->getSymRefTab()->findOrCreateArrayShadowSymbolRef(TR::Address, anchoredArrayBaseAddressNode);
+   TR::Node *elementLoadNode = TR::Node::createWithSymRef(comp->il.opCodeForIndirectArrayLoad(TR::Address), 1, 1, elementAddress, elementSymRef);
+   elementLoadNode->copyByteCodeInfo(node);
+
+   TR::TreeTop *elementLoadTT = NULL;
+   if (comp->useCompressedPointers())
+      elementLoadTT = originalBlock->append(TR::TreeTop::create(comp, TR::Node::createCompressedRefsAnchor(elementLoadNode)));
+   else
+      elementLoadTT = originalBlock->append(TR::TreeTop::create(comp, TR::Node::create(node, TR::treetop, 1, elementLoadNode)));
+
+   if (enableTrace)
+      traceMsg(comp, "Created array element load treetop n%dn node n%dn\n", elementLoadTT->getNode()->getGlobalIndex(), elementLoadNode->getGlobalIndex());
+
+
+   ///////////////////////////////////////
+   // 7. Store the return value from the array element load to the same register or temp used by the anchored node
+   TR::Node *storeArrayElementNode = createStoreNodeForAnchoredNode(anchoredNode, elementLoadNode, comp, enableTrace, "array element load");
+
+   elementLoadTT->insertAfter(TR::TreeTop::create(comp, storeArrayElementNode));
+
+   if (enableTrace)
+      traceMsg(comp, "Store array element load node n%dn to n%dn %s\n", elementLoadNode->getGlobalIndex(), storeArrayElementNode->getGlobalIndex(), storeArrayElementNode->getOpCode().getName());
+
+
+   ///////////////////////////////////////
+   // 8. Split (3) at the array element load node
+   TR::Block *arrayElementLoadBlock = originalBlock->split(elementLoadTT, cfg);
+
+   arrayElementLoadBlock->setIsExtensionOfPreviousBlock(true);
+
+   if (enableTrace)
+      traceMsg(comp, "Isolated array element load treetop n%dn node n%dn in block_%d\n", elementLoadTT->getNode()->getGlobalIndex(), elementLoadNode->getGlobalIndex(), arrayElementLoadBlock->getNumber());
+
+   int32_t dataWidth = TR::Symbol::convertTypeToSize(TR::Address);
+   if (comp->useCompressedPointers())
+      dataWidth = TR::Compiler->om.sizeofReferenceField();
+
+   TR::Node *arraylengthNode = TR::Node::create(TR::arraylength, 1, anchoredArrayBaseAddressNode);
+   arraylengthNode->setArrayStride(dataWidth);
+
+   elementLoadTT->insertBefore(TR::TreeTop::create(comp, TR::Node::createWithSymRef(TR::NULLCHK, 1, 1, arraylengthNode, comp->getSymRefTab()->findOrCreateNullCheckSymbolRef(comp->getMethodSymbol()))));
+
+   elementLoadTT->insertBefore(TR::TreeTop::create(comp, TR::Node::createWithSymRef(TR::BNDCHK, 2, 2, arraylengthNode, anchoredElementIndexNode, comp->getSymRefTab()->findOrCreateArrayBoundsCheckSymbolRef(comp->getMethodSymbol()))));
```

Comment:
I was just thinking again about this `BNDCHK` and the one generated for array element store, as well.  Would it make sense to generate it during IL generation so that it could be exposed to later optimizations?  Then any bound checking that's being performed in the helpers could be dropped.

I guess I'm not sure how much opportunity for improvement that might present.  Vijay @vijaysun-omr Daryl @0xdaryl, any thoughts?  I guess this is something we could evaluate in another round of improvements.

Reply:
@hzongaro that is a good question. Is there a reason that anyone can think of for why the bound check should not be generated in the IL generator ?

Reply:
> Then any bound checking that's being performed in the helpers could be dropped.

If `BNDCHK` is not applicable to the helper call case but generated in ILGen, when splitting the blocks in tree lowering, we have to make sure `BNDCHK` is split/moved into the same block as the regular aaload/aastore blocks. It adds an extra layer of complexity. If it really benefits, I think it could be done in another round of improvements as Henry mentioned.

Reply:
> If BNDCHK is not applicable to the helper call case but generated in ILGen

What I'm thinking is that we would make it applicable to both the helper call and the inline code, so it would stay in `originalBlock`, just as with the `NULLCHK`. . . .  But I'm OK with leaving it for now, and going back to this in another round, as it entails changes to IL generation, my pending pull request to Value Propagation, and the helpers to properly evaluate the benefit.

Reply:
I also agree. @hzongaro last comment (especially the interaction with VP change being delivered) captures my thinking very well.

Reply:
This enhancement work is documented in #12553